solr/core/src/java/org/apache/solr/managed/ResourceManagerPool.java

```diff
@@ -0,0 +1,294 @@
+    }
+  }
+
+  public Object setResourceLimit(T component, String limitName, Object value, ChangeListener.Reason reason) throws Exception {
+    try {
+      Object newActualLimit = doSetResourceLimit(component, limitName, value);
+      changeCounts.computeIfAbsent(reason, r -> new AtomicLong()).incrementAndGet();
+      for (ChangeListener listener : listeners) {
+        listener.changedLimit(getName(), component, limitName, value, newActualLimit, reason);
+      }
+      return newActualLimit;
+    } catch (Throwable t) {
+      for (ChangeListener listener : listeners) {
+        listener.onError(getName(), component, limitName, value, reason, t);
+      }
+      throw t;
+    }
+  }
+
+  protected abstract Object doSetResourceLimit(T component, String limitName, Object value) throws Exception;
+
+  public abstract Map<String, Object> getResourceLimits(T component) throws Exception;
+
+  /**
+   * Calculate aggregated monitored values.
+   * <p>Default implementation of this method simply sums up all non-negative numeric values across
+   * components and ignores any non-numeric values.</p>
+   */
+  public Map<String, Object> aggregateTotalValues(Map<String, Map<String, Object>> perComponentValues) {
+    // calculate the totals
+    Map<String, Object> newTotalValues = new HashMap<>();
+    perComponentValues.values().forEach(map -> map.forEach((k, v) -> {
+      // only calculate totals for numbers
+      if (!(v instanceof Number)) {
+        return;
+      }
+      Double val = ((Number)v).doubleValue();
+      // -1 and MAX_VALUE are our special guard values
+      if (val < 0 || val.longValue() == Long.MAX_VALUE || val.longValue() == Integer.MAX_VALUE) {
+        return;
+      }
+      newTotalValues.merge(k, val, (v1, v2) -> ((Number)v1).doubleValue() + ((Number)v2).doubleValue());
+    }));
+    return newTotalValues;
+  }
+
+  /** Get current pool limits. */
+  public Map<String, Object> getPoolLimits() {
+    return Collections.unmodifiableMap(poolLimits);
+  }
+
+  /**
+   * Pool limits are defined using controlled tags.
+   */
+  public void setPoolLimits(Map<String, Object> poolLimits) {
+    if (poolLimits != null) {
+      this.poolLimits = new HashMap(poolLimits);
+    }
+  }
+
+  /** Get parameters specified during creation. */
+  public Map<String, Object> getPoolParams() {
+    return Collections.unmodifiableMap(poolParams);
+  }
+
+  public void setPoolParams(Map<String, Object> poolParams) {
+    if (poolParams != null) {
+      this.poolParams = new HashMap<>(poolParams);
+      this.scheduleDelaySeconds = Integer.parseInt(String.valueOf(poolParams.getOrDefault(SCHEDULE_DELAY_SECONDS_PARAM, DEFAULT_SCHEDULE_DELAY_SECONDS)));
+    }
+  }
+
+  /**
+   * Pool context used for managing additional pool state.
+   */
+  public ResourcePoolContext getResourcePoolContext() {
+    return poolContext;
+  }
+
+  public void manage() {
+    manageLock.lock();
```

Comment:
I think we can get rid of the manageLock completely. The manage method is only called by the scheduled executor which guarantees that the effect of previous executions happens-before the subsequent ones.